src/test/scripts/functions/codegenalg/Algorithm_PCA.R

```diff
@@ -0,0 +1,108 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+#
+# This script performs Principal Component Analysis (PCA) on the given input data.
+#
+
+args <- commandArgs(TRUE)
+library("Matrix")
+
+A = readMM(paste(args[1], "A.mtx", sep=""));
+K = ncol(A);
+#ofmt = "CSV";
+projectData = 0;
+model = "";
+center = 0;
+scale = 0;
+output =  "/";
+
+evec_dominant = matrix(0, 1, 1);
+
+if (model != "") {
+  # reuse existing model to project data
+} else if (model == "") {
+    model = output;
+}
+
+  N = nrow(A);
+  D = ncol(A);
+
+  # perform z-scoring (centering and scaling)
+  if (center == 1) {
+    cm = colMean(A);
+    A  = A - matrix(1, nrow(A), 1) %*% cm;
+  }
+  if (scale == 1) {
+    cvars = (colSums(A^2));
+    if (center == 1){
+      cm = colMeans(A);
+      cvars = (cvars - N*(cm^2))/(N-1);
+    }
+    Azscored = A / sqrt(cvars);
+    A = Azscored;
+  }
+
+  # co-variance matrix
+  mu = matrix(1, nrow(A), 1) %*% colSums(A)/N;
+  C = (t(A) %*% A)/(N-1) - (N/(N-1))*t(mu) %*% mu;
+
+  # list data structure to mutli assign function values
+  list <- structure(NA,class="result")
+  "[<-.result" <- function(x,...,value) {
+    args <- as.list(match.call())
+    args <- args[-c(1:2,length(args))]
+    length(value) <- length(args)
+    for(i in seq(along=args)) {
+      a <- args[[i]]
+      if(!missing(a)) eval.parent(substitute(a <- v,list(a=a,v=value[[i]])))
+    }
+    x
+  }
+
+  # compute eigen vectors and values
+  list[evalues, evectors] = eigen(C);
+
+  decreasing_Idx = order(as.vector(evectors), decreasing=TRUE);
+  diagmat = table(seq(1,D), decreasing_Idx[1:D]);
+  # sorts eigen values by decreasing order
+  evalues = diagmat %*% evalues;
+  # sorts eigen vectors column-wise in the order of decreasing eigen values
+  evectors = evectors %*% diagmat;
+
+  # select K dominant eigen vectors
+  nvec = ncol(evectors);
+
+  eval_dominant = evalues[1:K, 1];
+  evec_dominant = evectors[,1:K];
+  Eps = 10^-8;
+  # the square root of eigen values
+  eval_stdev_dominant = sqrt(eval_dominant); # NaNs produced here!
```

Comment:
Hi @mboehm7 , `eval_dominant`, is producing `NaN`s.

Reply:
I would recommend to first simplify and fix the script: First, you can simply call R = eigen(C)
R <- eigen(C), and get the results via evectors = r$vectors and evalues <- R$values. So there is no need for your custom list function (which might potentially corrupt the results). Second, there is also a type regarding colMean vs colMeans. Third, please be careful with matrix-vector operation which are not supported in R - also colMeans does not return a row vector but a vector which is interpreted as row vector in matrix operations.